File: Java/src/AES.java
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.util.HashMap;

public class AES {
	public static void main(String cliArgs[]) {
		HashMap<String, String> args = Args.parse(cliArgs);
		HashMap<String, Object> io = IO.loadIO(args);
		AESKeys keys = new AESKeys((byte[]) io.get("key"));

		switch(args.get("act")) {
			case "e":
				encrypt(io, keys);
				break;
			case "d":
				decrypt(io, keys);
				break;
			case "h":
				hash(io, keys);
				break;
		}
		IO.unloadIO(io);
	};

	/*
	** block functions
	*/
	private static void encryptBlock(AESMatrix matrix, AESKeys keys) {
		matrix.addRoundKey(keys.get(0));

		for(int i = 1; i < 11; i++) {
			matrix.subBytes();
			matrix.shiftRows();
			if(i != 10){
				matrix.mixColumns();
			}
			matrix.addRoundKey(keys.get(i));
		}
	}

	private static void decryptBlock(AESMatrix matrix, AESKeys keys) {
		for(int i = 10; i > 0; i--) {
			matrix.addRoundKey(keys.get(i));
			if(i != 10){
				matrix.rmixColumns();
			}
			matrix.rshiftRows();
			matrix.rsubBytes();
		}

		matrix.addRoundKey(keys.get(0));
	}

	/*
	** Stream functions
	*/
	private static void encrypt(HashMap<String, Object> io, AESKeys keys) {
		byte[] rawMatrix = new byte[16];
		int read = IO.readBlock((FileInputStream) io.get("input"), rawMatrix);

		while(read > 0) {
			AESMatrix matrix = new AESMatrix(rawMatrix);
			encryptBlock(matrix, keys);
			try{
				matrix.writeToDisk((FileOutputStream) io.get("output"));
			} catch(Exception e) {
				System.out.println("Error: could not output the cypher");
				System.exit(-1);
			}
			if(read == 15) {
				rawMatrix = new byte[16]; // make it all zeroes
				rawMatrix[15] = 17;
				read = 16;
				continue;
			}

			read = IO.readBlock((FileInputStream) io.get("input"), rawMatrix);
		}
	}

	private static void decrypt(HashMap<String, Object> io, AESKeys keys) {
		long blocks = (long)io.get("i_size") / 16;
		boolean zeroFlag = false;
		byte[] rawMatrix = new byte[16];
		FileOutputStream output = (FileOutputStream) io.get("output");

		for(int i = 0; i < blocks; ++i) {
			IO.readBlock((FileInputStream) io.get("input"), rawMatrix);
			AESMatrix matrix = new AESMatrix(rawMatrix);
			decryptBlock(matrix, keys);

			try {

				if(i == blocks - 2 && matrix.get(15) == 0) {
					zeroFlag = true;
					matrix.writeToDisk(output, 15);
				} else if(i != blocks - 1) {
					matrix.writeToDisk(output);
				} else if(IO.allZeroes(matrix)) {
					if(matrix.get(15) == 17) {
						if(!zeroFlag) {
							matrix.writeToDisk(output);
						} // else dont write, it's padding
					} else {
						if(zeroFlag) {
							output.write(0x00);
						}
						matrix.writeToDisk(output, 16 - matrix.get(15));
					}
				} else {
					if(zeroFlag) {
						output.write(0x00);
					}
					matrix.writeToDisk(output);
				}

			} catch(Exception e) {
				System.out.println("Error: could not output the plain text");
				System.exit(-1);
			}
		}
	}

	private static void hash(HashMap<String, Object> io, AESKeys keys) {
		AESMatrix out = new AESMatrix();
		byte[] rawMatrix = new byte[16];
		int read = IO.readBlock((FileInputStream) io.get("input"), rawMatrix);

		while(read > 0) {
			AESMatrix matrix = new AESMatrix(rawMatrix);
			encryptBlock(matrix, keys);
			out.addRoundKey(matrix);

			if(read == 15) {
				rawMatrix = new byte[16]; // make it all zeroes
				rawMatrix[15] = 17;
				read = 16;
				continue;
			}

			read = IO.readBlock((FileInputStream) io.get("input"), rawMatrix);
		}

		System.out.println(out.toLineString());
	}
}

File: Lua/src/aes.lua
local AESKeys = require("src/aes_keys")
local AESMatrix = require("src/aes_matrix")

local function encrypt_matrix(matrix, keys)
	matrix:add_round_key(keys:get(1))
	for i = 2, #keys.keys do
		matrix:sub_bytes()
		matrix:shift_rows()
		if i ~= #keys.keys then matrix:mix_columns() end
		matrix:add_round_key(keys:get(i))
	end
end

local function decrypt_matrix(matrix, keys)
	for i = #keys.keys, 2, -1 do
		matrix:add_round_key(keys:get(i))
		if i ~= #keys.keys then matrix:rmix_columns() end
		matrix:rshift_rows()
		matrix:rsub_bytes()
	end
	matrix:add_round_key(keys:get(1))
end

local function encrypt(aes_io)
	local bytes, len = aes_io.read_block()

	while len > 0 do
		local matrix = AESMatrix:new()
		matrix:from_bytes(bytes)
		encrypt_matrix(matrix, aes_io.io.keys)
		aes_io.io.output:write(matrix:to_bytes())

		if len == 15 then
			bytes = ""
			for i = 1, 15 do bytes = bytes .. string.char(0x00) end
			bytes = bytes .. string.char(0x11) --17
			len = 16
		else
			bytes, len = aes_io.read_block()
		end
	end
end

--[[
	Padding hell explained:
	upon reaching the penultimate block the program sets a flag if the last block is 0
	(in case the padding was 17 long) and doesnt write it till later.
	and it writes the blocks as they go until it gets to the last one.
	It checks the padding pattern using the function and if it held now it has two options:
	one, the padding pattern is 17 (16 zeroes and flag is up), thus it discards it all,
	unless the flag is down then the pattern broke, so it writes all.
	If the pattern was sub 16, and the flag was up, that means the byte was held by mistake so it writes it,
	and then applies the substring adequately.
	Finally if the last block turns out not to be padded we make sure we didn't hold a zero off
	and write the whole thing.
]]
local function decrypt(aes_io)
	local blocks = aes_io.io.input:seek("end") / 16 -- size should be multiple of 16
	aes_io.io.input:seek("set") -- return to start
	local zero_flag = false --gets set when the penultimate block ends with 0

	for i = 1, blocks do
		local bytes, _ = aes_io.read_block()
		local matrix = AESMatrix:new()
		matrix:from_bytes(bytes)
		decrypt_matrix(matrix, aes_io.io.keys)

		-- dealing with padding hell sarts here
		if i == blocks - 1 and matrix.get(16) == 0x00 then
			zero_flag = true
			aes_io.io.output:write(matrix:to_bytes():sub(1,15))

		elseif i ~= blocks then
			aes_io.io.output:write(matrix:to_bytes())

		elseif aes_io.check_pad(matrix) then -- if pad detected
			if matrix.get(16) == 0x11 then
				if not zero_flag then
					aes_io.io.output:write(matrix:to_bytes())
				end -- if the flag was up we would discard the pad
			else
				if zero_flag then aes_io.io.output:write(string.char(0x00)) end
				aes_io.io.output:write(matrix:to_bytes():sub(1, 16 - matrix.get(16)))
			end

		else
			if zero_flag then aes_io.io.output:write(string.char(0x00)) end
			aes_io.io.output:write(matrix:to_bytes())
		end -- end of hell
	end
end

local function hash(aes_io)
	local result = AESMatrix:new()
	local bytes, len = aes_io.read_block()

	while len > 0 do
		local matrix = AESMatrix:new()
		matrix:from_bytes(bytes)
		encrypt_matrix(matrix, aes_io.io.keys)
		result:xor_aes_matrix(matrix)

		if len == 15 then
			bytes = ""
			for i = 1, 15 do bytes = bytes .. string.char(0x00) end
			bytes = bytes .. string.char(0x11) --17
			len = 16
		else
			bytes, len = aes_io.read_block()
		end
	end
	print(result:to_bytes():gsub(".", function(c)
		return string.format("%2x", c:byte())
	end))

end

-- return statement
return {
	hash = hash,
	encrypt = encrypt,
	decrypt = decrypt
}
